### ISSUE TITLE
bug: induction intros reverted assumptions even in goals that don’t result in the motive

### ISSUE THREAD

**Issue by iehality** [leanprover/lean4#3212]
### Prerequisites

* [x] Put an X between the brackets on this line if you have done all of the following:
    * Check that your issue is not already [filed](https://github.com/leanprover/lean4/issues).
    * Reduce the issue to a minimal, self-contained, reproducible test case. Avoid dependencies to mathlib4 or std4.

### Description

The induction tactic does not behave as intended when there are restrictions on motive.

### Context

[Zulip chat](https://leanprover.zulipchat.com/#narrow/stream/270676-lean4/topic/restricted.20induction)

Thanks to @nomeata for pointing out that this might be a bug.

### Steps to Reproduce

Execute the following code in web editor.
```lean
axiom N : Type

axiom zero : N

axiom succ : N → N

axiom add : N → N → N

axiom Restriction : (N → Prop) → Prop

axiom restricted_induction {motive : N → Prop} (h : Restriction motive)
  (zero : motive zero) (succ : ∀ n, motive n → motive (succ n)) : ∀ n, motive n

example (n m : N) (h : n ≠ zero) : add m n ≠ zero := by
  induction n using restricted_induction
```

**Expected behavior:**
 Three goals are generated: `Restriction fun n => n ≠ zero → add m n ≠ zero`, `add m 0 ≠ zero` and `add m (succ n) ≠ zero`. In later two, the hypothesis `h` are rewriten to `zero ≠ zero` and `succ n ≠ zero`, respectively.

**Actual behavior:** 
Returns following error.
```
tactic 'introN' failed, insufficient number of binders

nm: N
⊢ Restriction fun n => n ≠ zero → add m n ≠ zero
```

### Versions

4.5.0-rc1
Web editor

### Impact

Add :+1: to [issues you consider important](https://github.com/leanprover/lean4/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc). If others are impacted by this issue, please ask them to add :+1: to it.

**Comment by nomeata** [comment]
Thanks! Copying some analysis from Zulip here:

Looking at the code a bit, I one possible fix would be to extend docs#Lean.Meta.Match.Alt with a field indicating whether this “alternative” as the `motive` as the result type, and then use that field in docs#Lean.Elab.Tactic.ElimApp.evalAlts to run
```
          (_, altMVarId) ← altMVarId'.introNP numGeneralized
```
only if it is set. Probably worth fixing.

**Comment by iehality** [comment]
I think it is difficult for me due to my limited knowledge about metaprogramming.

**Comment by nomeata** [comment]
That’s fine. I’ll give it a shot after #3188.

**Comment by nomeata** [comment]
Came up again in https://leanprover.zulipchat.com/#narrow/stream/348111-std4/topic/Insufficient.20number.20of.20targets.

And I guess me working on this has been preemted. I better unassign, so if someone else wants to they dare to.

**Comment by nomeata** [comment]
Ok, fix ripening in #3256.
